File: clients/zig/sync/tests/test_un.zig
// Tests for the Zig unsandbox SDK
// Run with: zig test tests/test_un.zig

const std = @import("std");
const un = @import("../src/un.zig");
const testing = std.testing;
const mem = std.mem;

// ============================================================================
// Unit Tests - Test exported library functions
// ============================================================================

test "detectLanguage" {
    try testing.expectEqualStrings("python", un.detectLanguage("script.py") orelse "");
    try testing.expectEqualStrings("javascript", un.detectLanguage("script.js") orelse "");
    try testing.expectEqualStrings("typescript", un.detectLanguage("script.ts") orelse "");
    try testing.expectEqualStrings("go", un.detectLanguage("script.go") orelse "");
    try testing.expectEqualStrings("rust", un.detectLanguage("script.rs") orelse "");
    try testing.expectEqualStrings("c", un.detectLanguage("script.c") orelse "");
    try testing.expectEqualStrings("cpp", un.detectLanguage("script.cpp") orelse "");
    try testing.expectEqualStrings("d", un.detectLanguage("script.d") orelse "");
    try testing.expectEqualStrings("zig", un.detectLanguage("script.zig") orelse "");
    try testing.expectEqualStrings("bash", un.detectLanguage("script.sh") orelse "");
    try testing.expectEqualStrings("ruby", un.detectLanguage("script.rb") orelse "");
    try testing.expectEqualStrings("php", un.detectLanguage("script.php") orelse "");
    try testing.expect(un.detectLanguage("script.unknown") == null);
    try testing.expect(un.detectLanguage("script") == null);
}

test "version" {
    const v = un.version();
    try testing.expect(v.len > 0);
    // Should be in semver format (at least "0.0.0")
    try testing.expect(v.len >= 5);
}

test "lastError" {
    // Set an error
    un.setLastError("test error message");

    // Retrieve it
    const err = un.lastError();
    try testing.expectEqualStrings("test error message", err);

    // Clear it
    un.setLastError("");
    const err2 = un.lastError();
    try testing.expectEqualStrings("", err2);
}

test "SDK_VERSION constant" {
    try testing.expect(un.SDK_VERSION.len > 0);
    try testing.expect(un.SDK_VERSION.len >= 5);
}

// ============================================================================
// Integration Tests - Test SDK internal consistency
// ============================================================================

test "hmacSign" {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const secret_key = "test-secret";
    const message = "test-message";

    const result = try un.hmacSign(allocator, secret_key, message);
    defer allocator.free(result);

    // Should return a 64-character hex string
    try testing.expectEqual(@as(usize, 64), result.len);

    // Should be deterministic
    const result2 = try un.hmacSign(allocator, secret_key, message);
    defer allocator.free(result2);
    try testing.expectEqualStrings(result, result2);

    // Different inputs should produce different outputs
    const result3 = try un.hmacSign(allocator, secret_key, "different-message");
    defer allocator.free(result3);
    try testing.expect(!mem.eql(u8, result, result3));
}

// ============================================================================
// Functional Tests - Test against real API (requires credentials)
// These tests are marked as skipped by default since they require credentials
// ============================================================================

fn hasCredentials() bool {
    const pk = std.posix.getenv("UNSANDBOX_PUBLIC_KEY") orelse return false;
    const sk = std.posix.getenv("UNSANDBOX_SECRET_KEY") orelse return false;
    return pk.len > 0 and sk.len > 0;
}

fn getCredentials() ?struct { pk: []const u8, sk: []const u8 } {
    const pk = std.posix.getenv("UNSANDBOX_PUBLIC_KEY") orelse return null;
    const sk = std.posix.getenv("UNSANDBOX_SECRET_KEY") orelse return null;
    if (pk.len == 0 or sk.len == 0) return null;
    return .{ .pk = pk, .sk = sk };
}

test "healthCheck functional" {
    if (!hasCredentials()) {
        std.debug.print("SKIP (no credentials)\n", .{});
        return;
    }

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const healthy = try un.healthCheck(allocator);
    std.debug.print("Health check result: {}\n", .{healthy});
}

test "getLanguages functional" {
    const creds = getCredentials() orelse {
        std.debug.print("SKIP (no credentials)\n", .{});
        return;
    };

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const result = try un.getLanguages(allocator, creds.pk, creds.sk);
    defer allocator.free(result);

    try testing.expect(result.len > 0);
    // Should contain python
    try testing.expect(mem.indexOf(u8, result, "python") != null);
}

test "validateKeys functional" {
    const creds = getCredentials() orelse {
        std.debug.print("SKIP (no credentials)\n", .{});
        return;
    };

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const result = try un.validateKeys(allocator, creds.pk, creds.sk);
    defer allocator.free(result);

    try testing.expect(result.len > 0);
}

test "sessionList functional" {
    const creds = getCredentials() orelse {
        std.debug.print("SKIP (no credentials)\n", .{});
        return;
    };

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const result = try un.sessionList(allocator, creds.pk, creds.sk);
    defer allocator.free(result);

    try testing.expect(result.len > 0);
}

test "serviceList functional" {
    const creds = getCredentials() orelse {
        std.debug.print("SKIP (no credentials)\n", .{});
        return;
    };

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const result = try un.serviceList(allocator, creds.pk, creds.sk);
    defer allocator.free(result);

    try testing.expect(result.len > 0);
}

test "snapshotList functional" {
    const creds = getCredentials() orelse {
        std.debug.print("SKIP (no credentials)\n", .{});
        return;
    };

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const result = try un.snapshotList(allocator, creds.pk, creds.sk);
    defer allocator.free(result);

    try testing.expect(result.len > 0);
}

test "imageList functional" {
    const creds = getCredentials() orelse {
        std.debug.print("SKIP (no credentials)\n", .{});
        return;
    };

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const result = try un.imageList(allocator, null, creds.pk, creds.sk);
    defer allocator.free(result);

    try testing.expect(result.len > 0);
}

File: clients/cpp/sync/tests/test_un.cpp
// Tests for the C++ unsandbox SDK
// Compile: g++ -std=c++17 -o test_un test_un.cpp -I../src
// Run: ./test_un

#include <iostream>
#include <cassert>
#include <cstring>
#include <cstdlib>

// Include the SDK source directly for testing
// In production, you'd link against the compiled library
#include "../src/un.cpp"

using namespace std;

int tests_passed = 0;
int tests_failed = 0;

#define TEST(name) void test_##name()
#define RUN_TEST(name) do { \
    cout << "Running " << #name << "..." << endl; \
    try { \
        test_##name(); \
        cout << "  PASS" << endl; \
        tests_passed++; \
    } catch (const exception& e) { \
        cout << "  FAIL: " << e.what() << endl; \
        tests_failed++; \
    } catch (...) { \
        cout << "  FAIL: Unknown exception" << endl; \
        tests_failed++; \
    } \
} while(0)

#define ASSERT(cond) do { \
    if (!(cond)) { \
        throw runtime_error("Assertion failed: " #cond); \
    } \
} while(0)

#define ASSERT_EQ(a, b) do { \
    if ((a) != (b)) { \
        throw runtime_error("Assertion failed: " #a " == " #b); \
    } \
} while(0)

#define ASSERT_NE(a, b) do { \
    if ((a) == (b)) { \
        throw runtime_error("Assertion failed: " #a " != " #b); \
    } \
} while(0)

// ============================================================================
// Unit Tests - Test exported library functions
// ============================================================================

TEST(detect_language) {
    ASSERT_EQ(detect_language("script.py"), "python");
    ASSERT_EQ(detect_language("script.js"), "javascript");
    ASSERT_EQ(detect_language("script.ts"), "typescript");
    ASSERT_EQ(detect_language("script.go"), "go");
    ASSERT_EQ(detect_language("script.rs"), "rust");
    ASSERT_EQ(detect_language("script.c"), "c");
    ASSERT_EQ(detect_language("script.cpp"), "cpp");
    ASSERT_EQ(detect_language("script.d"), "d");
    ASSERT_EQ(detect_language("script.zig"), "zig");
    ASSERT_EQ(detect_language("script.sh"), "bash");
    ASSERT_EQ(detect_language("script.lua"), "lua");
    ASSERT_EQ(detect_language("script.php"), "php");
    ASSERT_EQ(detect_language("script.unknown"), "");
    ASSERT_EQ(detect_language("script"), "");
}

TEST(hmac_sign) {
    string secret_key = "test-secret";
    string message = "test-message";

    string result = hmac_sign(secret_key, message);

    // Should return a 64-character hex string
    ASSERT_EQ(result.length(), 64u);

    // Should be deterministic
    string result2 = hmac_sign(secret_key, message);
    ASSERT_EQ(result, result2);

    // Different inputs should produce different outputs
    string result3 = hmac_sign(secret_key, "different-message");
    ASSERT_NE(result, result3);
}

TEST(version) {
    string v = version();
    ASSERT(!v.empty());
    // Should be in semver format (at least "0.0.0")
    ASSERT(v.length() >= 5);
}

TEST(last_error) {
    // Set an error
    set_last_error("test error message");

    // Retrieve it
    string err = last_error();
    ASSERT_EQ(err, "test error message");

    // Clear it
    set_last_error("");
    err = last_error();
    ASSERT(err.empty());
}

TEST(escape_json) {
    ASSERT_EQ(escape_json("hello"), "hello");
    ASSERT_EQ(escape_json("hello\"world"), "hello\\\"world");
    ASSERT_EQ(escape_json("line1\nline2"), "line1\\nline2");
    ASSERT_EQ(escape_json("tab\there"), "tab\\there");
    ASSERT_EQ(escape_json("back\\slash"), "back\\\\slash");
}

TEST(base64_encode) {
    ASSERT_EQ(base64_encode(""), "");
    ASSERT_EQ(base64_encode("f"), "Zg==");
    ASSERT_EQ(base64_encode("fo"), "Zm8=");
    ASSERT_EQ(base64_encode("foo"), "Zm9v");
    ASSERT_EQ(base64_encode("foob"), "Zm9vYg==");
    ASSERT_EQ(base64_encode("fooba"), "Zm9vYmE=");
    ASSERT_EQ(base64_encode("foobar"), "Zm9vYmFy");
}

// ============================================================================
// Integration Tests - Test SDK internal consistency
// ============================================================================

TEST(compute_hmac) {
    string key = "test-key";
    string msg = "test-message";

    string sig1 = compute_hmac(key, msg);
    string sig2 = compute_hmac(key, msg);

    // Should be deterministic
    ASSERT_EQ(sig1, sig2);

    // Should produce different results for different inputs
    string sig3 = compute_hmac(key, "different");
    ASSERT_NE(sig1, sig3);
}

TEST(build_auth_headers) {
    string pk = "unsb-pk-test-test-test-test";
    string sk = "unsb-sk-test1-test2-test3-test4";

    string headers = build_auth_headers("POST", "/execute", "{}", pk, sk);

    // Should contain auth header
    ASSERT(headers.find("Authorization: Bearer " + pk) != string::npos);
    // Should contain timestamp header
    ASSERT(headers.find("X-Timestamp:") != string::npos);
    // Should contain signature header
    ASSERT(headers.find("X-Signature:") != string::npos);
}

// ============================================================================
// Functional Tests - Test against real API (requires credentials)
// ============================================================================

bool has_credentials() {
    const char* pk = getenv("UNSANDBOX_PUBLIC_KEY");
    const char* sk = getenv("UNSANDBOX_SECRET_KEY");
    return pk != nullptr && sk != nullptr && strlen(pk) > 0 && strlen(sk) > 0;
}

TEST(health_check_functional) {
    if (!has_credentials()) {
        cout << "  SKIP (no credentials)" << endl;
        return;
    }

    bool healthy = health_check();
    // Just verify it doesn't crash
    cout << "  Health check result: " << (healthy ? "healthy" : "unhealthy") << endl;
}

TEST(get_languages_functional) {
    if (!has_credentials()) {
        cout << "  SKIP (no credentials)" << endl;
        return;
    }

    const char* pk = getenv("UNSANDBOX_PUBLIC_KEY");
    const char* sk = getenv("UNSANDBOX_SECRET_KEY");

    string result = get_languages(pk, sk);
    ASSERT(!result.empty());
    // Should contain python
    ASSERT(result.find("python") != string::npos);
}

TEST(validate_keys_functional) {
    if (!has_credentials()) {
        cout << "  SKIP (no credentials)" << endl;
        return;
    }

    const char* pk = getenv("UNSANDBOX_PUBLIC_KEY");
    const char* sk = getenv("UNSANDBOX_SECRET_KEY");

    string result = validate_keys(pk, sk);
    ASSERT(!result.empty());
}

TEST(execute_functional) {
    if (!has_credentials()) {
        cout << "  SKIP (no credentials)" << endl;
        return;
    }

    const char* pk = getenv("UNSANDBOX_PUBLIC_KEY");
    const char* sk = getenv("UNSANDBOX_SECRET_KEY");

    string result = execute("python", "print('hello from cpp test')", pk, sk);
    ASSERT(!result.empty());
    // Should contain output
    ASSERT(result.find("stdout") != string::npos || result.find("output") != string::npos);
}

TEST(session_list_functional) {
    if (!has_credentials()) {
        cout << "  SKIP (no credentials)" << endl;
        return;
    }

    const char* pk = getenv("UNSANDBOX_PUBLIC_KEY");
    const char* sk = getenv("UNSANDBOX_SECRET_KEY");

    string result = session_list(pk, sk);
    ASSERT(!result.empty());
}

TEST(service_list_functional) {
    if (!has_credentials()) {
        cout << "  SKIP (no credentials)" << endl;
        return;
    }

    const char* pk = getenv("UNSANDBOX_PUBLIC_KEY");
    const char* sk = getenv("UNSANDBOX_SECRET_KEY");

    string result = service_list(pk, sk);
    ASSERT(!result.empty());
}

TEST(snapshot_list_functional) {
    if (!has_credentials()) {
        cout << "  SKIP (no credentials)" << endl;
        return;
    }

    const char* pk = getenv("UNSANDBOX_PUBLIC_KEY");
    const char* sk = getenv("UNSANDBOX_SECRET_KEY");

    string result = snapshot_list(pk, sk);
    ASSERT(!result.empty());
}

TEST(image_list_functional) {
    if (!has_credentials()) {
        cout << "  SKIP (no credentials)" << endl;
        return;
    }

    const char* pk = getenv("UNSANDBOX_PUBLIC_KEY");
    const char* sk = getenv("UNSANDBOX_SECRET_KEY");

    string result = image_list("", pk, sk);
    ASSERT(!result.empty());
}

int main() {
    cout << "===== C++ SDK Tests =====" << endl << endl;

    // Unit tests
    cout << "--- Unit Tests ---" << endl;
    RUN_TEST(detect_language);
    RUN_TEST(hmac_sign);
    RUN_TEST(version);
    RUN_TEST(last_error);
    RUN_TEST(escape_json);
    RUN_TEST(base64_encode);

    cout << endl << "--- Integration Tests ---" << endl;
    RUN_TEST(compute_hmac);
    RUN_TEST(build_auth_headers);

    cout << endl << "--- Functional Tests ---" << endl;
    RUN_TEST(health_check_functional);
    RUN_TEST(get_languages_functional);
    RUN_TEST(validate_keys_functional);
    RUN_TEST(execute_functional);
    RUN_TEST(session_list_functional);
    RUN_TEST(service_list_functional);
    RUN_TEST(snapshot_list_functional);
    RUN_TEST(image_list_functional);

    cout << endl << "===== Results =====" << endl;
    cout << "Passed: " << tests_passed << endl;
    cout << "Failed: " << tests_failed << endl;

    return tests_failed > 0 ? 1 : 0;
}
